### ISSUE TITLE
Expose `mprotect` in mmap

### ISSUE THREAD

**Issue by ronaldoussoren** [python/cpython#114233]
In #113868 I noted that we currently don't expose `mprotect` functionality in `mmap` and that this might be useful to have for use cases like generating native code in Python.  @brandtbucher confirmed that this would be useful to have

>
>            > * There is no way to change the memory protection (`PROT_READ` etc.) after creating the mapping, exposing such functionality could be useful for using `mmap` to write executable code in Python.
>
> Sorry I'm late, but +1 for exposing `mprotect`. I was prototyping JITs in pure-Python a while back, and it was unfortunate to have to hack around the edges with `ctypes` just to do something like this (which "feels like" it should be part of the `mmap` module).
>
> _Originally posted by @brandtbucher in https://github.com/python/cpython/issues/113868#issuecomment-1894557101_
      
Adding this should be fairly easy, given that the `mmap` code assumes that the entire buffer has the same protections.  One thing to consider is the impact on sharing memory through the buffer interface. 

### Linked PRs
* gh-114674

**Comment by brandtbucher** [comment]
Note to whoever implements this: we probably want to flush the i-cache whenever the executable bit is added to a buffer (or an already-executable buffer is written to). Or else expose this functionality to the user somehow.

**Comment by ronaldoussoren** [comment]
Good point.  GCC, clang and msvc have builtin functions for this, YMMV on other compilers.  A separate API for flushing the cache is probably best given that protections might be W+X, possibly with automatically calling that when a buffer is made executable. 

To make life more fun, macOS is slightly different: https://developer.apple.com/documentation/apple-silicon/porting-just-in-time-compilers-to-apple-silicon?language=objc, and https://developer.apple.com/forums/thread/672804. The former describes how to implements JIT on macOS (involving different APIs than mprotect), the latter says that there are problems with using mprotect with MAP_JIT (at least in macOS 11.2).

**Comment by brandtbucher** [comment]
> To make life more fun, macOS is slightly different: https://developer.apple.com/documentation/apple-silicon/porting-just-in-time-compilers-to-apple-silicon?language=objc, and https://developer.apple.com/forums/thread/672804. The former describes how to implements JIT on macOS (involving different APIs than mprotect), the latter says that there are problems with using mprotect with MAP_JIT (at least in macOS 11.2).

We were just discussing this over on #113465. According to some experiments I ran locally (which seem to agree with the forum thread above), the `MAP_JIT` flag and `pthread_jit_write_protect_np` calls are only necessary (a) on AArch64 macOS and (b) when the region is both `PROT_WRITE` and `PROT_EXEC`. Otherwise, the region can be `mmap`'d and `mprotect`'d as either `PROT_WRITE` xor `PROT_EXEC` without any shenanigans.

I think the cache-clearing stuff is only *needed* on ARM, too, since (I think) Intel has better cache coherency. But it wouldn't hurt, since the compiler builtins for doing it just compile away on those platforms.

Not sure how any of this changes with signed/packaged Python executables, though. Any special insight there?

**Comment by aisk** [comment]
I tried to add the `mprotect` method in #114674. But for the i-cache flush, do you mean by calling the compiler built-in clear_cache? 

If we want to expose it, how about adding a new method called `clear_cache`, or just calling the `clear_cache` in the flush method if the executable flag is set?

**Comment by ronaldoussoren** [comment]
We first need to work out the API, in particular if it is necessary to support different protection modes on subranges of a mmap buffer.  The current mmap code assumes that the protection mode is the same for the entire buffer.

Also, the correct way to do this (JIT)  on macOS is different than just using mprotect, see the link I shared earlier. Any API needs to accommodate that.  I haven't formed an opinion yet on how such an API should look.

**Comment by aisk** [comment]
Hi @ronaldoussoren, I don't think too much about adding the `start` / `length` parameters to this new method, just noticed that the existing `madvise` method already has these parameters.

I re-checked the current implementation and found that there is an `access` field in `mmap_object`, which indicates the current access mode of the mmap object. Therefore, I agree with you that we shouldn't expose the ability to let users change only parts of the object's access mode.

However, I also just realized that the current constructor has two ways to create a mmap object: using the Python-defined `access` flag or the more low-level `prot`/`flag` pair. So, for this new method, do we need to support both ways of specifying the protection modes? Please note that on Windows, only the `access` flag is allowed for protection modes in the constructor.

**Comment by nullhook** [comment]
how about `mmap(exec=true)`?
